client/src/instance_spec/mod.rs

```diff
@@ -0,0 +1,224 @@
+    Crucible { gen: u64, serialized_req: String },
+
+    /// A device backed by a file on the host machine. The payload is a path to
+    /// this file.
+    File { path: String },
+
+    /// A device backed by an in-memory buffer in the VMM process. The payload
+    /// is the buffer's contents.
+    InMemory { bytes: Vec<u8> },
+}
+
+/// A storage backend.
+#[derive(Clone, Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct StorageBackend {
+    /// The kind of storage backend this is.
+    pub kind: StorageBackendKind,
+
+    /// Whether the storage is read-only.
+    pub readonly: bool,
+}
+
+/// A kind of storage device: the sort of virtual device interface the VMM
+/// exposes to guest software.
+#[derive(Clone, Copy, Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub enum StorageDeviceKind {
+    Virtio,
+    Nvme,
+}
+
+/// A storage device.
+#[derive(Clone, Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct StorageDevice {
+    /// The device interface to present to the guest.
+    pub kind: StorageDeviceKind,
+
+    /// The name of the device's backend.
+    pub backend_name: String,
+
+    /// The PCI path at which to attach this device.
+    pub pci_path: PciPath,
+}
+
+//
+// Network devices.
+//
+// Because there is only one kind of virtual networking device (a virtio-net
+// device with an enlightened vNIC), networking devices don't have the same
+// BackendKind and DeviceKind enums that storage devices do, but they can
+// be added if needed in future versions.
+
+/// A network backend, specifically a virtual NIC on the host system for which
+/// Virtio network adapter integration is enabled.
+#[derive(Clone, Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct NetworkBackend {
+    /// The name of the vNIC to connect to.
+    pub vnic_name: String,
+}
+
+/// A virtual network adapter that presents a virtio network device interface.
+#[derive(Clone, Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct NetworkDevice {
+    /// The name of the device's backend.
+    pub backend_name: String,
+
+    /// The PCI path at which to attach this device.
+    pub pci_path: PciPath,
+}
+
+//
+// Serial ports.
+//
+
+/// A serial port identifier, which determines what I/O ports a guest can use to
+/// access a port.
+#[derive(Clone, Copy, Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub enum SerialPortNumber {
+    Com1,
+    Com2,
+    Com3,
+    Com4,
+}
+
+/// A serial port device.
+#[derive(Clone, Copy, Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct SerialPort {
+    /// The serial port number for this port.
+    pub num: SerialPortNumber,
+
+    /// The initial discard discipline for this port. If true, the device will
+    /// not buffer bytes written from the guest until the port's discipline
+    /// changes.
+    pub autodiscard: bool,
```

Comment:
Perhaps something for later down the line, but I'm not certain we should be exposing this in such a manner.  Serial ports with no connected backend can probably be auto-discard by default.  Those with a connected backend can let it decide. (I suspect?)

Reply:
Done--this makes it easier to reason about whether two serial ports are compatible for migration purposes, so I think it makes sense just to do this now.